src/api/general-services/pipeline-manage/constructors/create-new-step.js

```diff
@@ -0,0 +1,91 @@
+const config = require('../../../../config');
+
+const createNewStep = (context, step, args) => {
+  const {
+    processingConfig, clusterInfo, experimentId, pipelineImages, accountId,
+  } = context;
+
+  const { taskName } = args;
+
+
+  const task = JSON.stringify({
+    experimentId,
+    taskName,
+    config: processingConfig[taskName] || {},
+  });
+
+  if (config.clusterEnv === 'development') {
+    return {
+      ...step,
+      Type: 'Task',
+      Resource: 'arn:aws:states:::lambda:invoke',
+      Parameters: {
+        FunctionName: `arn:aws:lambda:eu-west-1:${accountId}:function:local-container-launcher`,
+        Payload: {
+          image: pipelineImages['remoter-client'],
+          name: 'pipeline-remoter-client',
+          task,
+          detached: false,
+        },
+      },
+    };
+  }
+
+
+  return {
+    ...step,
+    Type: 'Task',
+    Comment: 'Attempts to create a Kubernetes Job for the pipeline server. Will swallow a 409 (already exists) error.',
+    Resource: 'arn:aws:states:::eks:call',
+    Parameters: {
+      ClusterName: clusterInfo.name,
+      CertificateAuthority: clusterInfo.certAuthority,
+      Endpoint: clusterInfo.endpoint,
+      Method: 'POST',
+      Path: `/apis/batch/v1/namespaces/${config.workerNamespace}/jobs`,
+      RequestBody: {
+        apiVersion: 'batch/v1',
+        kind: 'Job',
+        metadata: {
+          name: `remoter-client-${experimentId}`,
+        },
+        spec: {
+          template: {
+            metadata: {
+              name: `remoter-client-${experimentId}`,
+            },
+            spec: {
+              containers: [
```

Comment:
doesn't a lot of this overlap with the object that `createNewJobIfNotExist` returns? Have you checked if it makes sense to extract the common logic in one place? Or is it going to lead to leaky abstraction?

Reply:
Actually, great point Iva.

The leaky abstraction point is not going to be an issue imho. This is becausethe `createNewJobIfNotExist` function's only purpose is to *create* a server. It does not monitor its lifecycle as it's meant to be run long-term.

For this constructor (`create-new-step`), there is a proper way to run a Job and wait for it to finish -- this is documented [here](https://docs.aws.amazon.com/step-functions/latest/dg/connect-eks.html). Instead of using `arn:aws:states::eks:call` and just create a Job like we do with the server (which will run throughout the pipeline), the steps themselves should be executed using `arn:aws:states:::eks:runJob.sync`. This makes it possible for the state machine to wait for the Job to exit successfully or with an error and stop the machine if there has been an error.

We should use the special-purpose `arn:aws:states:::eks:runJob.sync` for `create-new-step.js` accordingly.

Reply:
Changed in commit 344f4bf